Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp

```diff
@@ -0,0 +1,192 @@
+            "to %d%% with moving average).", originalBufferingPercentage, correctedBufferingPercentage2, extraElements,
+            averagedBufferingPercentage);
+    }
+#endif
+
+    return averagedBufferingPercentage;
+}
+
+void GStreamerQuirkBroadcomBase::resetBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, int bufferingPercentage) const
+{
+    if (!isEnsuredOwnedState(playerPrivate))
+        return;
+
+    GStreamerQuirkBroadcomBaseState& state = static_cast<GStreamerQuirkBroadcomBaseState&>(playerPrivate->quirkState());
+
+    state.m_streamBufferingLevelMovingAverage.reset(bufferingPercentage);
+}
+
+void GStreamerQuirkBroadcomBase::setupBufferingPercentageCorrection(MediaPlayerPrivateGStreamer* playerPrivate, GstState currentState, GstState newState, GstElement* element) const
+{
+    if (!isEnsuredOwnedState(playerPrivate))
+        return;
+
+    GStreamerQuirkBroadcomBaseState& state = static_cast<GStreamerQuirkBroadcomBaseState&>(playerPrivate->quirkState());
+
+    // This code must support being run from different GStreamerQuirkBroadcomBase subclasses without breaking. Only the
+    // first subclass instance should run.
+    if (currentState == GST_STATE_NULL && newState == GST_STATE_READY
+        && g_strstr_len(GST_ELEMENT_NAME(element), 13, "brcmvidfilter")) {
+        state.m_vidfilter = element;
+
+        // Also get the multiqueue (if there's one) attached to the vidfilter. We'll need it later to correct the buffering level.
+        for (auto* sinkPad : GstIteratorAdaptor<GstPad>(GUniquePtr<GstIterator>(gst_element_iterate_sink_pads(state.m_vidfilter.get())))) {
+            GRefPtr<GstPad> peerSrcPad = adoptGRef(gst_pad_get_peer(sinkPad));
+            if (peerSrcPad) {
+                GRefPtr<GstElement> peerElement = adoptGRef(GST_ELEMENT(gst_pad_get_parent(peerSrcPad.get())));
+                // The multiqueue reference is useless if we can't access its stats (on older GStreamer versions).
+                if (peerElement && g_strstr_len(GST_ELEMENT_NAME(peerElement.get()), 10, "multiqueue")
+                    && g_object_class_find_property(G_OBJECT_GET_CLASS(peerElement.get()), "stats"))
+                    state.m_multiqueue = peerElement;
+            }
+            break;
+        }
+    } else if (currentState == GST_STATE_NULL && newState == GST_STATE_READY
+        && g_strstr_len(GST_ELEMENT_NAME(element), 6, "queue2")) {
+        state.m_queue2 = element;
+    } else if (currentState == GST_STATE_READY && newState == GST_STATE_NULL
+        && g_strstr_len(GST_ELEMENT_NAME(element), 13, "brcmvidfilter")) {
+        state.m_vidfilter = nullptr;
+        state.m_multiqueue = nullptr;
+    } else if (currentState == GST_STATE_READY && newState == GST_STATE_NULL
+        && g_strstr_len(GST_ELEMENT_NAME(element), 6, "queue2")) {
+        state.m_queue2 = nullptr;
+    }
+}
+
+bool GStreamerQuirkBroadcomBase::isEnsuredOwnedState(MediaPlayerPrivateGStreamer* playerPrivate) const
```

Comment:
I understand what you're asking for and I also thought about that approach (one state per quirk, multiple quirks, all of them held by the player in a map) **but**, in this specific case of actions performed by the GStreamerQuirkBroadcomBase superclass of GStreamerQuirkBroadcom and GStreamerQuirkBcmNexus, that approach poses a problem:

All these buffering percentage corrections are "accumulative"/"destructive". They take the original percentage, ask some data to some extra pipeline elements, update a moving average and return the "decorated" percentage. In a system that needs both the GStreamerQuirkBroadcom and GStreamerQuirkBcmNexus quirks, letting the two of them "correct" the percentage would mean that the original percentage would be corrected (increased/decreased) twice, and that's wrong. Both quirks are competing for the same correction, and only **one** of them must do it (no matter which one as long as one does it). That, on systems that have only the BcmNexus quirk, on systems that have only the Broadcom quirk, and on systems that have both.

The solution that I could come with, stores a single state, belonging to a single quirk (the first one who attempts to configure the state on the player) and disallows the other competing quirks to operate on that state. This ensures that at most one quirk performs the correction.

I agree that storing multiple states, one per quirk, would be interesting in the future for those use cases, just not for this specific one.

Reply:
From what I understand, this correction is done in the base class so it does not matter if it's performed by any of the instantiated subclasses. Then, when running the quirk, you should perform just one correction and bail out when that correction is done. I understand you are already doing that at `GStreamerQuirksManager::correctBufferingPercentage` as you perform the correction and then return.

If that were not enough because you might correct the percertage and still get the same result, which would force the second correction, then that method signature should be changed to maybe return an `std::optional<int>` for example and the moment you get a valid value, you bail out.

With all this in mind, I think you can keep my suggestion without convoluting the structures in such a way you proposed to fix this small inconvenience. Am I wrong?

Reply:
You are right about not correcting twice by using an optional<int>, but I'm not so happy about all the subclasses setting up the state and having to keep references to several elements. If only one subclass should actually operate on each player, why wasting space keeping a state for the idle subclass (idle as in "lost the competition to be in charge of correcting the percentage)?

Maybe I can come up with a way to support multiple states, but prevent the last subclass that arrives to the party against configuring itself and operating. It just feels like a lot of overengineering for a problem that should actually be simpler.

Reply:
Maybe you can do something, yes, but I think we are trying to over-engineer here to save just an idle structure. A compromise here could be trying the change the signature of the hash map to take an ASCIILiteral as key and then the base class could install the structure, which could be the same for both subclasses. That way we install the structures we need when we need them.

Reply:
Ok. As we discussed in person, I'm going to change all the methods in GStreamerQuirk related to buffering percentage correction to tell in some way if they have processed the request or not. That way, the loops in GStreamerQuirksManager that ask all the registered quirks can bail out as soon as one of the quirks attends the request (returns true or returns a value on it s std::optional). This will effectively prevent more than one GStreamerQuirkBroadcomBase subclass instance to be in charge of buffering level correction, as intended.